src/mlpack/tests/main_tests/hmm_viterbi_test.cpp

```diff
@@ -0,0 +1,166 @@
+    RandomInitialize(hmm.Emission());
+  }
+
+  //! Helper function to create discrete HMM.
+  static void Create(HMM<DiscreteDistribution>& hmm,
+                     vector<mat>& trainSeq,
+                     size_t states,
+                     double tolerance = 1e-05)
+  {
+    // Maximum observation is necessary so we know how to train the discrete
+    // distribution.
+    arma::Col<size_t> maxEmissions(trainSeq[0].n_rows);
+    maxEmissions.zeros();
+    for (vector<mat>::iterator it = trainSeq.begin(); it != trainSeq.end();
+         ++it)
+    {
+      arma::Col<size_t> maxSeqs =
+          arma::conv_to<arma::Col<size_t>>::from(arma::max(*it, 1)) + 1;
+      maxEmissions = arma::max(maxEmissions, maxSeqs);
+    }
+
+    hmm = HMM<DiscreteDistribution>(size_t(states),
+        DiscreteDistribution(maxEmissions), tolerance);
+  }
+
+  static void Create(HMM<GaussianDistribution>& hmm,
+                     vector<mat>& trainSeq,
+                     size_t states,
+                     double tolerance = 1e-05)
+  {
+    // Not implemented
+    // Prevent unused parameter warning
+    (void)hmm;
+    (void)trainSeq;
+    (void)states;
+    (void)tolerance;
+  }
+
+  static void Create(HMM<GMM>& hmm,
+                     vector<mat>& trainSeq,
+                     size_t states,
+                     double tolerance = 1e-05)
+  {
+    // Not implemented
+    // Prevent unused parameter warning
+    (void)hmm;
+    (void)trainSeq;
+    (void)states;
+    (void)tolerance;
+  }
+
+  //! Helper function for discrete emission distributions.
+  static void RandomInitialize(vector<DiscreteDistribution>& e)
+  {
+    for (size_t i = 0; i < e.size(); ++i)
+    {
+      e[i].Probabilities().randu();
+      e[i].Probabilities() /= arma::accu(e[i].Probabilities());
+    }
+  }
+
+  static void RandomInitialize(vector<GaussianDistribution>& e)
+  {
+    // Not implemented
+    // Prevent unused parameter warning
+    (void)e;
+  }
+
+  static void RandomInitialize(vector<GMM>& e)
+  {
+    // Not implemented
+    // Prevent unused parameter warning
+    (void)e;
+  }
+};
+
+struct Train
+{
+  template<typename HMMType>
+  static void Apply(HMMType& hmm, vector<arma::mat>* trainSeq)
+  {
+    // For now, perform unsupervised (Baum-Welch) training
+    hmm.Train(*trainSeq);
+  }
+};
+
+BOOST_AUTO_TEST_CASE(HMMViterbiCheckDimensionsTest)
+{
+  // Train an HMM
+  HMMModel * h = new HMMModel(DiscreteHMM);
+  // Load data
+  arma::mat inp;
+  data::Load("obs1.csv", inp);
+  std::vector<arma::mat> trainSeq = {inp};
+  // Init
+  h->PerformAction<Init, std::vector<arma::mat>>(&trainSeq);
+  // Train
+  h->PerformAction<Train, std::vector<arma::mat>>(&trainSeq);
+
+  // Set the params for the hmm_viterbi invocation
+  SetInputParam("input_model", h);
+  SetInputParam("input", inp);
+
+  mlpackMain();
+
+  arma::Mat<size_t> out = CLI::GetParam<arma::Mat<size_t> >("output");
+  BOOST_REQUIRE(out.n_rows == inp.n_rows);
```

Comment:
Are you sure this is right?  I thought `output` should have one row containing the states computed by the Viterbi alignment.  (Maybe `inp.n_rows` is 1, but still, we should probably change the test to be more clear about this.  This would also mean the `n_elem` check isn't necessary.)